app/models/attachment_reference.rb

```diff
@@ -0,0 +1,65 @@
+class AttachmentReference < ActiveRecord::Base
+  after_initialize :set_expiry_at, if: :attachable_nil?
+  before_validation :update_expiry_at
+
+  belongs_to :attachable, polymorphic: true, inverse_of: :attachment_references
+  belongs_to :attachment, inverse_of: :attachment_references
+
+  delegate :open, :url, :path, to: :attachment
+
+  # Get the name from the file and then further build or find an attachment based on file's SHA256
+  # hash.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  def file=(file)
+    self.name = filename(file)
+    self.attachment = Attachment.find_or_initialize_by(file: file)
+  end
+
+  # Clear the expiry time when attachable is present and set the expiry time when attachable is
+  # removed.
+  def attachable=(attachable)
```

Comment:
@lowjoel Debugged again, `attachable=` is not called when you do a save from the attachable object. code like `FactoryGirl.create(:material)` won't clear the expiry time.